File: 84.largest-rectangle-in-histogram.java
import java.util.Stack;

/*
 * @lc app=leetcode id=84 lang=java
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
class Solution {
    /**
     * 暴力法
     * 
     * @param heights
     * @return
     */
    public int largestRectangleArea1(int[] heights) {
        int max = 0;
        for (int i = 0; i < heights.length; ++i) {
            for (int j = i; j < heights.length; ++j) {
                int minHeight = Integer.MAX_VALUE;
                for (int k = i; k <= j; ++k) {
                    minHeight = Math.min(minHeight, heights[k]);
                }
                max = Math.max(max, minHeight * (j - i + 1));
            }
        }
        return max;
    }

    /**
     * 辅助栈法
     * 
     * @param heights
     * @return
     */
    public int largestRectangleArea2(int[] heights) {
        int max = 0;
        Stack<Rect> stack = new Stack<>();
        stack.push(new Rect(-1, -1));
        for (int i = 0; i < heights.length; ++i) {
            while (stack.peek().height >= heights[i]){
                max = Math.max(max, stack.pop().height * (i - stack.peek().position - 1));
            }
            stack.push(new Rect(i, heights[i]));
            // if (heights[i] >= stack.peek().height) {
            //     stack.push(new Rect(i, heights[i]));
            // } else {
            //     max = Math.max(max, stack.pop().height * (i - stack.peek().position - 1));
            // }
        }
        while (stack.size() > 1) {
            Rect rect = stack.pop();
            max = Math.max(max, rect.height * (heights.length - stack.peek().position - 1));
        }
        return max;
    }

    class Rect {
        int position = 0;
        int height = 0;

        public Rect(int position, int height){
            this.position = position;
            this.height = height;
        };
    }
}
// @lc code=end

File: 84.largest-rectangle-in-histogram.py
#
# @lc app=leetcode id=84 lang=python3
#
# [84] Largest Rectangle in Histogram
#
# https://leetcode.com/problems/largest-rectangle-in-histogram/description/
#
# algorithms
# Hard (33.88%)
# Likes:    3093
# Dislikes: 77
# Total Accepted:    237.2K
# Total Submissions: 699.9K
# Testcase Example:  '[2,1,5,6,2,3]'
#
# Given n non-negative integers representing the histogram's bar height where
# the width of each bar is 1, find the area of largest rectangle in the
# histogram.
# 
# 
# 
# 
# Above is a histogram where width of each bar is 1, given height =
# [2,1,5,6,2,3].
# 
# 
# 
# 
# The largest rectangle is shown in the shaded area, which has area = 10
# unit.
# 
# 
# 
# Example:
# 
# 
# Input: [2,1,5,6,2,3]
# Output: 10
# 
# 
#

# @lc code=start
class Solution:
    # python超时
    def largestRectangleArea(self, heights: List[int]) -> int:
        ans = 0
        for i in range(len(heights)):
            left, right = i, i
            while left > 0:
                left -= 1
                if heights[left] < heights[i]:
                    left += 1
                    break
            while right < len(heights) - 1:
                right += 1
                if heights[right] < heights[i]:
                    right -= 1
                    break
            ans = max(ans, heights[i] * (right - left + 1))
        return ans
            
        
# @lc code=end

